### DISCUSSION TITLE
classes/class_enetconnection

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#84] [User-contributed notes]
# classes/class_enetconnection

Inherits: RefCounted< Object A wrapper class for an ENetHost. Description: ENet's purpose is to provide a relatively thin, simple and robust network communication layer on top of UDP (User Datagram...

https://docs.godotengine.org/en/stable/classes/class_enetconnection.html

**Comment by filipkrw** [comment]
Let me save you some time. Upon receiving an `EVENT_RECEIVE` event from `service` method, the packet (event data) will not be in the the returned `Array`. To get it, you need to call `get_packet` on the associated peer:

```cpp
Array result = conn->service(0);
Ref<ENetPacketPeer> peer = result[1];
PackedByteArray data = peer->get_packet();
```

**Comment by Poraksis** [comment]
This auto genarated documents assume that i know everything. I want to learn build my game client and server separately without high level api. As i understood i need ENetConnection for both client and server but to determine server side logic over client i need to filter in service right? Btw ENetPackedPeer is a complete binding so i may try to follow ENet from other sources

**Comment by Elrinth** [comment]
Is ENET possible to use in a browser for joining or hosting a game? Because for me, I'm getting failed to connect to server if I'm running in browser and hosting on same pc.

```
WARNING: Sending failed!
 at: enet_socket_send (thirdparty/enet/godot.cpp:528)
```

Same when trying to host in browser: 

```
ERROR: Couldn't create an ENet host.
   at: _create (modules/enet/enet_connection.cpp:318)
ERROR: Supplied MultiplayerPeer must be connecting or connected.
   at: set_multiplayer_peer (modules/multiplayer/scene_multiplayer.cpp:196)"
```

**Reply by Calinou** [reply]
> Is ENET possible to use in a browser for joining or hosting a game?

No, because [web browsers can't send raw UDP packets](https://www.gafferongames.com/post/why_cant_i_send_udp_packets_from_a_browser/) (and ENet is a layer on top of UDP). Use WebSockets or WebRTC instead. WebSockets has higher latency but is easy to set up, while WebRTC can provide latency similar to ENet but is significantly more involved to set up.
